keps/sig-auth/2579-psp-replacement/README.md

```diff
@@ -0,0 +1,969 @@
+
+Migrating to the replacement policy from PodSecurityPolicies can be done effectively using a
+combination of dry-run and audit/warn modes (although this becomes harder if mutating PSPs are
+used).
+
+We could also ship a standalone tool to assist with the PodSecurityPolicy migration. Here are some
+ideas for the sorts of things the tool could assist with:
+
+- Analyze PSP resources, identify the closest profile level, and highlight the differences
+- Check the authorization mode for existing pods. For example, if a pod’s service account is not
+  authorized to use the PSP that validated it (based on the `kubernetes.io/psp` annotation), then
+  that should trigger a warning.
+- Automate the dry-run and/or labeling process.
+- Automatically select (and optionally apply) a policy level for each namespace.
+
+We should also publish a step-by-step migration guide. A rough approach might look something like
+this, with the items tagged (automated) having support from the PSP migration tool.
+
+1. Enable 3-tier policy admission plugin, default everything to privileged.
+2. Eliminate mutating PSPs:
+    1. Clone all mutating PSPs to a non-mutating version (automated)
+    2. Update all ClusterRoles authorizing use of the mutating PSPs to also authorize use of the
+       non-mutating variant (automated)
+    3. Watch for pods using the mutating PSPs (check via the `kubernetes.io/psp` annotation), and
+       work with code owners to migrate to valid non-mutating resources.
+    4. Delete mutating PSPs
+3. Select a compatible profile for each namespace, based on the existing resources in the namespace
+   (automated)
+    1. Review the profile choices
+    2. Evaluate the difference in privileges that would come from disabling the PSP controller
+       (automated).
+4. (optional) Apply the profiles in `warn` and `audit` mode (automated)
+5. Apply the profiles in `allow` mode (automated)
+6. Disable PodSecurityPolicy
+
+### Custom Profiles
+
+Allow custom profile levels to be statically configured. E.g.
+`--extra-pod-security-levels=host-network`. Custom profiles are ignored by the built-in admission
+plugin, and must be handled completely by a 3rd party webhook (including the dry-run implementation,
+if desired).
+
+### Custom Warning Messages
+
+An optional `prefix/warn-message` annotation can be used to return a custom warning message (in
+addition to the standard message) whenever a policy warning is triggered. This could be useful for
+announcing the date that new policy will take effect, or providing a point of contact if you need to
+request an exception.
+
+### Windows Support
+
+Even without built-in support, enforcement for Windows pods can be delegated to a webhook admission
+plugin by exempting the `windows` RuntimeClass. We should investigate built-in Windows support out
+of the box though. Requirements include:
+
+1. A standardized identifier for Windows pods
+2. Write the Pod Security Standards for windows
```

Comment:
we have details of which policies we would want to cover from a windows perspective: https://github.com/kubernetes/kubernetes/issues/64801#issuecomment-668897952.  We were waiting till we had implementation (we've started this in gatekeeper) before drafting the policy so we didn't give guidance and have no way to apply the guidance.

Reply:
Would the idea for adding Windows support be to add an additional `plugin` to the `AdmissionConfiguration`? or a separate `AdmissionConfiguration`?

The relevant PSP for windows are the following. These would be the same recomendations as restricted/privileged/baseline on Linux.  There will be some details in implementation like running as Nonroot would need to check for `ContainerAdministrator` instead of root user. This is done in kubelet [today](https://github.com/kubernetes/kubernetes/pull/92355).

- Usage of volume types
- Usage of the host filesystem
- Running as Non-root

There are on windows specific one that should be added:

- GMSA

With Windows [HostProcess](https://github.com/containerd/containerd/pull/5131/files) there would be additional check required.